atom/browser/api/atom_api_update_client.cc

```diff
@@ -0,0 +1,172 @@
+      Emit("crx-installer-done");
+      break;
+    }
+  }
+}
+
+void UpdateClient::OnEvent(Events event, const std::string& id) {
+  update_client::CrxUpdateItem item;
+  GetComponentDetails(id, &item);
+  switch (event) {
+    case Events::COMPONENT_CHECKING_FOR_UPDATES:
+      Emit("component-checking-for-updates", id);
+      break;
+    case Events::COMPONENT_WAIT:
+      Emit("component-wait", id);
+      break;
+    case Events::COMPONENT_UPDATE_FOUND:
+      Emit("component-update-found", id);
+      break;
+    case Events::COMPONENT_UPDATE_READY: {
+      // Uninstall for upgrades
+        atom::api::Extension::GetInstance()->Remove(id);
+        Emit("component-update-ready", id);
+      }
+      break;
+    case Events::COMPONENT_UPDATED: {
+        auto str = item.component.version.GetString();
+        Emit("component-update-updated", id, str);
+      }
+      break;
+    case Events::COMPONENT_NOT_UPDATED:
+      Emit("component-not-updated", id);
+      break;
+    case Events::COMPONENT_UPDATE_DOWNLOADING:
+      Emit("component-update-downloading", id);
+      break;
+  }
+}
+
+void UpdateClient::RegisterComponentForUpdate(
+    const std::string& public_key,
+    const base::Closure& registered_callback,
+    const ReadyCallback& ready_callback) {
+  component_updater::RegisterExtension(
+      cus_.get(), public_key, registered_callback, ready_callback);
+}
+
+void UpdateClient::OnComponentRegistered(const std::string& extension_id) {
+  Emit("component-registered", extension_id);
+}
+
+void UpdateClient::OnComponentReady(
+    const std::string& extension_id,
+    const base::FilePath& install_dir) {
+  Emit("component-ready", extension_id, install_dir.value());
+}
+
+void UpdateClient::RegisterComponent(const std::string& extension_id) {
+  base::Closure registered_callback =
+    base::Bind(&UpdateClient::OnComponentRegistered,
+               base::Unretained(this), extension_id);
+  ReadyCallback ready_callback =
+    base::Bind(&UpdateClient::OnComponentReady,
+               base::Unretained(this), extension_id);
+  if (extension_id == kOnePasswordId) {
+    RegisterComponentForUpdate(
+        kOnePasswordPublicKeyStr, registered_callback, ready_callback);
+  } else if (extension_id == kDashlaneId) {
+    RegisterComponentForUpdate(
+        kDashlanePublicKeyStr, registered_callback, ready_callback);
+  } else if (extension_id == kLastPassId) {
+    RegisterComponentForUpdate(
+        kLastPassPublicKeyStr, registered_callback, ready_callback);
+  } else if (extension_id == kPDFJSId) {
+    RegisterComponentForUpdate(
+        kPDFJSPublicKeyStr, registered_callback, ready_callback);
+  }
+}
+
+void UpdateClient::CheckNow(const std::string& extension_id) {
+  OnDemandUpdate(cus_.get(), extension_id);
+}
+
+// static
+mate::Handle<UpdateClient> UpdateClient::Create(
+    v8::Isolate* isolate,
+    content::BrowserContext* browser_context) {
+  return mate::CreateHandle(isolate,
```

Comment:
the update client should use `chrome::GetBrowserContextRedirectedInIncognito` to make sure that there is only one per user since the extensions are shared across private and session contexts